app/services/contributors_generator.rb

```diff
@@ -0,0 +1,243 @@
+# typed: true
+# frozen_string_literal: true
+
+# generates Cocina::Models::Contributors to be used by DescriptionGenerator (ultimately in a Cocina::Models::RequestDRO)
+# rubocop:disable Metrics/ClassLength
+class ContributorsGenerator
+  extend T::Sig
+
+  sig { params(work: Work).returns(T::Array[Cocina::Models::Contributor]) }
+  def self.generate(work:)
+    new(work: work).generate
+  end
+
+  sig { params(work: Work).returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def self.form_from_contributors(work:)
+    new(work: work).form_from_contributors
+  end
+
+  sig { params(work: Work).void }
+  def initialize(work:)
+    @work = work
+  end
+
+  sig { returns(T::Array[Cocina::Models::Contributor]) }
+  def generate
+    result = []
+
+    work.contributors.each do |work_form_contributor|
+      result << contributor(work_form_contributor)
+    end
+
+    result
+  end
+
+  FORM_EVENT = Cocina::Models::DescriptiveValue.new(
+    {
+      value: 'Event',
+      type: 'resource types',
+      source: { value: 'DataCite resource types' }
+    }
+  )
+
+  sig { returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def form_from_contributors
+    work.contributors.each do |contributor|
+      next unless contributor.role == 'Conference' || contributor.role == 'Event'
+
+      return [FORM_EVENT]
+    end
+
+    []
+  end
+
+  private
+
+  sig { returns(Work) }
+  attr_reader :work
+
+  sig { params(contributor: Contributor).returns(T.nilable(Cocina::Models::Contributor)) }
+  def contributor(contributor)
+    # FIXME: TODO: mappings for status (primary) and/or order
+    Cocina::Models::Contributor.new(
+      "name": name_descriptive_value(contributor),
+      "type": contributor_type(contributor),
+      "role": cocina_roles(contributor.role)
+    )
+  end
+
+  sig { params(contributor: Contributor).returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def name_descriptive_value(contributor)
+    return [Cocina::Models::DescriptiveValue.new(value: full_name(contributor))] unless contributor.person?
+
+    [Cocina::Models::DescriptiveValue.new(value: full_name(contributor), type: 'inverted full name')]
+  end
+
+  sig { params(contributor: Contributor).returns(T.nilable(String)) }
+  def full_name(contributor)
+    return "#{contributor.last_name}, #{contributor.first_name}" if contributor.person?
+
+    contributor.full_name
+  end
+
+  sig { params(contributor: Contributor).returns(String) }
+  def contributor_type(contributor)
+    return 'conference' if contributor.role == 'Conference'
+    return 'event' if contributor.role == 'Event'
+
+    contributor.contributor_type
```

Comment:
So this method returns one of four values:

1. person
2. organization
3. conference
4. event

:question:

Reply:
so far, yes.